### PR TITLE
Mbs 7969 k8soidc

### PR DESCRIPTION
**Pull request by dsvoronin**

### CODE REVIEW

File: docs/content/docs/test/Runner.md [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+1. Get access to kubernetes cloud: [internal doc](http://links.k.avito.ru/Kubectl)
+1. [Request](http://links.k.avito.ru/androidEmulatorServiceDesk) `exec` access to `android-emulator` namespace in `beta` cluster
+1. Setup a context on `beta/android-emulator` with your user access (also described in first doc)
+1. Run tests with extra parameters specified, example for `:test-app`:
+
+```shell script
+./gradlew :subprojects:android-test:test-app:instrumentationUi 
+    -Pci=true 
+    -PkubernetesContext=<your context>
+    -PkubernetesCaCertFile=<kube cert>
```
Comment (sergioboi) [inline docs/content/docs/test/Runner.md]: Maybe we need an example path to file?

File: docs/content/docs/test/Runner.md [hunk 2]
```diff
@@ -0,0 +1,21 @@
+---
+title: Test Runner
+type: docs
+---
+
```
Comment (sergioboi) [inline docs/content/docs/test/Runner.md:5]: Add specific internal tag. @eugene-krivobokov help me, please.

File: docs/content/docs/test/Runner.md [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+# Avito test runner
+
+## How to run tests in kubernetes from a local machine (internal)
+
+1. Get access to kubernetes cloud: [internal doc](http://links.k.avito.ru/Kubectl)
+1. [Request](http://links.k.avito.ru/androidEmulatorServiceDesk) `exec` access to `android-emulator` namespace in `beta` cluster
+1. Setup a context on `beta/android-emulator` with your user access (also described in first doc)
```
Comment (sergioboi) [inline docs/content/docs/test/Runner.md]:
```suggestion
1. Setup a context on `beta/android-emulator` with your user access. Direct path to section with context setup
```

File: docs/content/docs/test/Runner.md [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+1. Run tests with extra parameters specified, example for `:test-app`:
+
+```shell script
+./gradlew :subprojects:android-test:test-app:instrumentationUi 
+    -Pci=true 
+    -PkubernetesContext=<your context>
```
Comment (sergioboi) [inline docs/content/docs/test/Runner.md:20]: It's not clear what context I need to type here.

File: docs/content/docs/test/Runner.md (hunk 2 above)
Comment (eugene-krivobokov) [reply docs/content/docs/test/Runner.md:5]: `{{< avito >}}`

https://avito-tech.github.io/avito-android/docs/contributing/docs/#internal-links

File: subprojects/gradle/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/reservation/client/kubernetes/KubernetesReservationClient.kt [hunk 5]
```diff
@@ -355,6 +357,8 @@ class KubernetesReservationClient(
         val deploymentSpecificationsMatchLabels = deploymentMatchLabels
             .plus("deploymentName" to deploymentName)
 
+        logger.info("deploymentSpecificationsMatchLabels = $deploymentSpecificationsMatchLabels")
```
Comment (sergioboi) [inline subprojects/gradle/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/reservation/client/kubernetes/KubernetesReservationClient.kt]: to debug level

File: subprojects/gradle/kubernetes/src/main/kotlin/com/avito/utils/gradle/KubernetesCredentials.kt [hunk 6]
```diff
@@ -1,18 +1,38 @@
...
+        return if (context.isNullOrBlank()) {
+            //todo it should be mandatory only in avito ci
+            // move it to instrumentationPlugin config to control
+            KubernetesCredentials.Service(
+                token = getMandatoryStringProperty("kubernetesToken"),
+                caCertData = getMandatoryStringProperty("kubernetesCaCertData"),
+                url = getMandatoryStringProperty("kubernetesUrl")
+            )
+        } else {
+            val caCertFile = getOptionalStringProperty("kubernetesCaCertFile", nullIfBlank = true)
```
Comment (sergioboi) [inline subprojects/gradle/kubernetes/src/main/kotlin/com/avito/utils/gradle/KubernetesCredentials.kt:39]: Why is it optional? Is `kubernetesContext` not enough for separate logic?

File: subprojects/gradle/kubernetes/src/main/kotlin/com/avito/utils/gradle/KubernetesCredentials.kt [hunk 7]
```diff
@@ -1,18 +1,38 @@
...
 val Project.kubernetesCredentials: KubernetesCredentials
-    get() = KubernetesCredentials(
-        token = this.getMandatoryStringProperty("kubernetesToken"),
-        caCertData = this.getMandatoryStringProperty("kubernetesCaCertData"),
-        url = this.getMandatoryStringProperty("kubernetesUrl")
-    )
+    get() {
+        val context = getOptionalStringProperty("kubernetesContext", nullIfBlank = true)
+        return if (context.isNullOrBlank()) {
+            //todo it should be mandatory only in avito ci
```
Comment (sergioboi) [inline subprojects/gradle/kubernetes/src/main/kotlin/com/avito/utils/gradle/KubernetesCredentials.kt]: Maybe it should mandatory only with ci=true or kuberenetes.enabled=true

File: docs/content/docs/test/Runner.md (hunk 1 above)
Comment (dsvoronin) [reply docs/content/docs/test/Runner.md]: wait what? ) like `/home/my.crt`? what could be unclear here?

File: docs/content/docs/test/Runner.md (hunk 2 above)
Comment (dsvoronin) [reply docs/content/docs/test/Runner.md:5]: thanks, never used it

File: docs/content/docs/test/Runner.md (hunk 4 above)
Comment (dsvoronin) [reply docs/content/docs/test/Runner.md:20]: it is described in list above

File: subprojects/gradle/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/reservation/client/kubernetes/KubernetesReservationClient.kt (hunk 5 above)
Comment (dsvoronin) [reply subprojects/gradle/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/reservation/client/kubernetes/KubernetesReservationClient.kt]: i'll remove this line completely

File: docs/content/docs/test/Runner.md (hunk 4 above)
Comment (dsvoronin) [reply docs/content/docs/test/Runner.md:20]: and you need to know about https://kubernetes.io/docs/tasks/access-application-cluster/configure-access-multiple-clusters/#define-clusters-users-and-contexts
beforehand

File: subprojects/gradle/kubernetes/src/main/kotlin/com/avito/utils/gradle/KubernetesCredentials.kt (hunk 6 above)
Comment (dsvoronin) [reply subprojects/gradle/kubernetes/src/main/kotlin/com/avito/utils/gradle/KubernetesCredentials.kt:39]: because custom certificate is optional, by default its data parsed from kube/config as plain bytes

File: subprojects/gradle/kubernetes/src/main/kotlin/com/avito/utils/gradle/KubernetesCredentials.kt (hunk 7 above)
Comment (dsvoronin) [reply subprojects/gradle/kubernetes/src/main/kotlin/com/avito/utils/gradle/KubernetesCredentials.kt]: it's out of the scope of this PR i think, a'll make a proposal of how i see this thing in future

i can remove todo block if you want

Comment (dsvoronin) [reply subprojects/gradle/kubernetes/src/main/kotlin/com/avito/utils/gradle/KubernetesCredentials.kt]: removed

File: docs/content/docs/test/Runner.md (hunk 1 above)
Comment (dsvoronin) [reply docs/content/docs/test/Runner.md]: change to
`    -PkubernetesCaCertFile=<path to ca certificate file>`

File: subprojects/gradle/cd/build.gradle.kts [hunk 8]
```diff
@@ -11,7 +11,6 @@ dependencies {
     implementation(project(":subprojects:gradle:android"))
     implementation(project(":subprojects:gradle:impact-shared"))
     implementation(project(":subprojects:gradle:impact"))
-    implementation(project(":subprojects:gradle:docs"))
```
Comment (eugene-krivobokov) [inline subprojects/gradle/cd/build.gradle.kts:14]: this change doesn't relate to the issue
it would be easier to make in a separate PR

File: docs/content/docs/test/Runner.md (hunk 1 above)
Comment (sergioboi) [reply docs/content/docs/test/Runner.md]: I supposed that user can't understand what is a cacert file and where he can find it on local computer

File: subprojects/gradle/kubernetes/src/main/kotlin/com/avito/utils/gradle/KubernetesCredentials.kt [hunk 9]
```diff
@@ -1,18 +1,36 @@
...
+    data class Service(
+        val token: String,
+        val caCertData: String,
+        val url: String
+    ) : KubernetesCredentials(), Serializable
+
+    data class Config(
+        val context: String,
+        val caCertFile: String? = null,
+        val configFile: String = "${System.getProperty("user.home")}/.kube/config"
```
Comment (eugene-krivobokov) [inline subprojects/gradle/kubernetes/src/main/kotlin/com/avito/utils/gradle/KubernetesCredentials.kt]: It's good to support `KUBECONFIG` env also.

Comment (sergioboi) [reply subprojects/gradle/kubernetes/src/main/kotlin/com/avito/utils/gradle/KubernetesCredentials.kt]: requireNonNull

File: subprojects/gradle/cd/build.gradle.kts (hunk 8 above)
Comment (dsvoronin) [reply subprojects/gradle/cd/build.gradle.kts:14]: but it blocks kubernetes upgrade

Comment (eugene-krivobokov) [reply subprojects/gradle/cd/build.gradle.kts:14]: I see

**Review by eugene-krivobokov — APPROVED** [review]
